### PR TITLE
Backtrack rewrite

### PR DESCRIPTION
**Pull request by BenCat07**
Bases on #986.

Changes Include:
- Completely remake backtrack
- Make Backtrack Object Oriented
- Make Getting ticks Way easier
- Fix conflicts with crithack
- Use a more reliable way to calculate ticks and good ticks
- Fix Backtrack Aimbot
- Improve Autobackstab
- Improve Triggerbot
- Fix no-lerp issues
- Integrate into Demoknight Warp
- Better Chams On Backtrack

I May have missed some things.

### CODE REVIEW

**Comment by BenCat07** [comment]
the TF2C compile error is seemingly random...

File: .gitignore [hunk 1]
```diff
@@ -289,3 +289,5 @@ scripts/updater-preferences
 !/modules/CMakeLists.txt
 
 *.autosave
+
+CMakeCacheBackup.txt
```
Comment (TotallyNotElite) [inline .gitignore:293]: Remove this. Fixed on my branch.

File: src/hacks/Aimbot.cpp [hunk 2]
```diff
@@ -1275,24 +1232,39 @@ int BestHitbox(CachedEntity *target)
...
+
+            if (data)
+            {
+                if (*backtrackVischeckAll)
+                    for (int j = 0; j < 18; j++)
+                    {
+                        if (IsEntityVectorVisible(target, (*data).hitboxes[j].center))
+                            return j;
+                    }
+                else if (IsEntityVectorVisible(target, (*data).hitboxes[0].center))
```
Comment (TotallyNotElite) [inline src/hacks/Aimbot.cpp:1249]: Explicitly use HEAD here. Maybe set type of J to the enum as well.

File: src/hacks/Aimbot.cpp [hunk 3]
```diff
@@ -1275,24 +1232,39 @@ int BestHitbox(CachedEntity *target)
...
-        if (IsBacktracking() && !projectile_mode && good_tick.first != -1)
+        // Else attempt to find any hitbox at all
+        if (shouldBacktrack())
         {
-            namespace bt = hacks::shared::backtrack;
-            for (int i = 0; i < 18; i++)
-                if (IsEntityVectorVisible(target, bt::headPositions[target->m_IDX][good_tick.first].hitboxes.at(i).center))
-                    return i;
+            // We already vischecked
+            if (!*backtrackVischeckAll)
```
Comment (TotallyNotElite) [inline src/hacks/Aimbot.cpp:1260]: So, even if you don't want to vischeck all hitboxes, you still do it?

File: src/hacks/Backtrack.cpp [hunk 4]
```diff
@@ -1,284 +1,310 @@
...
-    if (!shouldBacktrack())
+    // Update enabled status
+    isBacktrackEnabled = isEnabled();
+    if (!isBacktrackEnabled)
+    {
+        latency_rampup = 0.0f;
+        return;
+    }
+
+    // Only return here if the local player hasn't spawned in yet (so ping does not go down when dead)
```
Comment (TotallyNotElite) [inline src/hacks/Backtrack.cpp:153]: This does not check alive state at all, what do you mean??

File: src/hacks/Backtrack.cpp [hunk 5]
```diff
@@ -290,203 +316,245 @@ static void Draw()
...
+}
+
+// This function is so other files can Easily get the best tick matching their criteria
+std::optional<BacktrackData> getBestTick(CachedEntity *ent, std::function<bool(CachedEntity *ent, BacktrackData &data, std::optional<BacktrackData> &best_tick)> callback)
+{
+    std::optional<BacktrackData> best_tick;
+
+    // No data recorded
+    if (ent->m_IDX <= 0 || backtrack_data.size() < ent->m_IDX || !backtrack_data.at(ent->m_IDX - 1))
+        return best_tick;
```
Comment (TotallyNotElite) [inline src/hacks/Backtrack.cpp:457]: Return std::nullopt, wtf?

File: src/hacks/Backtrack.cpp [hunk 6]
```diff
@@ -290,203 +316,245 @@ static void Draw()
...
-    return true;
+    // How big a difference is between the ping we fake and the tick passed?
+    int delta_ticks = current_tickcount - TIME_TO_TICKS(getLatency() / 1000.0f) - tick.tickcount;
+
+    // Difference may not be greater than 200ms, shave off a few ticks just to be safe.
+    if (fabsf(TICKS_TO_TIME(delta_ticks)) <= 0.2f - TICKS_TO_TIME(3))
+        return true;
+
+    return false;
+}
```
Comment (TotallyNotElite) [inline src/hacks/Backtrack.cpp:429]: You should probably just use MrSteyk's method for determining a good tick, which takes lerp etc into consideration.

File: src/hacks/Backtrack.cpp [hunk 7]
```diff
@@ -290,203 +316,245 @@ static void Draw()
...
 
-void EmptyBacktrackData(BacktrackData &i)
+bool defaultEntFilter(CachedEntity *ent)
 {
-    i = {};
+    // Dormant
+    if (CE_BAD(ent))
+        return false;
+    // Friend check
+    if (playerlist::IsFriend(ent))
```
Comment (TotallyNotElite) [inline src/hacks/Backtrack.cpp:485]: The year is 2020 and yet you still fucked this up xd

https://github.com/nullworks/cathook/blob/5b9af83d910e93ab95475059edce55b0a786ed93/src/PlayerTools.cpp#L42 is what you're looking for

File: src/hacks/Backtrack.cpp (hunk 4 above)
Comment (BenCat07) [reply src/hacks/Backtrack.cpp:153]: Yeah i removed that but forgot to edit the comment.

File: src/hacks/Backtrack.cpp (hunk 6 above)
Comment (BenCat07) [reply src/hacks/Backtrack.cpp:429]: This does too.

File: src/hacks/Backtrack.cpp (hunk 7 above)
Comment (BenCat07) [reply src/hacks/Backtrack.cpp:485]: oh god damn it

File: src/hacks/Aimbot.cpp (hunk 2 above)
Comment (BenCat07) [reply src/hacks/Aimbot.cpp:1249]: For loop with enum does not work out well

Aimbot.cpp:1244:58: error: cannot increment expression of enum type 'hitbox_t'

File: src/hacks/Aimbot.cpp (hunk 3 above)
Comment (BenCat07) [reply src/hacks/Aimbot.cpp:1260]: Wait why did i do this

File: src/hacks/Aimbot.cpp (hunk 2 above)
Comment (BenCat07) [reply src/hacks/Aimbot.cpp:1249]: i still made it look better. but yeah.

**Pull request merged**